pkg/autoupdate/tuf/library_manager.go

```diff
@@ -0,0 +1,341 @@
+		return nil
+	}
+
+	stagedUpdatePath, err := ulm.stageUpdate(binary, targetFilename)
+	defer func() {
+		if err := os.Remove(stagedUpdatePath); err != nil {
+			level.Debug(ulm.logger).Log("msg", "could not remove staged update", "staged_update_path", stagedUpdatePath, "err", err)
+		}
+	}()
+	if err != nil {
+		return fmt.Errorf("could not stage update: %w", err)
+	}
+
+	if err := ulm.verifyStagedUpdate(binary, stagedUpdatePath); err != nil {
+		return fmt.Errorf("could not verify staged update: %w", err)
+	}
+
+	if err := ulm.moveVerifiedUpdate(binary, targetFilename, stagedUpdatePath); err != nil {
+		return fmt.Errorf("could not move verified update: %w", err)
+	}
+
+	if currentVersion != nil {
+		ulm.tidyLibrary(binary, currentVersion)
+	} else {
+		level.Debug(ulm.logger).Log("msg", "skipping tidying library because current running version could not be determined", "binary", binary)
+	}
+
+	return nil
+}
+
+// alreadyAdded checks if the given target already exists in the update library.
+func (ulm *updateLibraryManager) alreadyAdded(binary string, targetFilename string) bool {
+	updateDirectory := filepath.Join(ulm.updatesDirectory(binary), ulm.versionFromTarget(binary, targetFilename))
+
+	return autoupdate.CheckExecutable(context.TODO(), executableLocation(updateDirectory, binary), "--version") == nil
+}
+
+// versionFromTarget extracts the semantic version for an update from its filename.
+func (ulm *updateLibraryManager) versionFromTarget(binary string, targetFilename string) string {
+	// The target is in the form `launcher-0.13.6.tar.gz` -- trim the prefix and the file extension to return the version
+	prefixToTrim := fmt.Sprintf("%s-", binary)
+
+	return strings.TrimSuffix(strings.TrimPrefix(targetFilename, prefixToTrim), ".tar.gz")
+}
+
+// stageUpdate downloads the update indicated by `targetFilename` and stages it for
+// further verification.
+func (ulm *updateLibraryManager) stageUpdate(binary string, targetFilename string) (string, error) {
+	stagedUpdatePath := filepath.Join(ulm.stagedUpdatesDirectory(binary), targetFilename)
+	out, err := os.Create(stagedUpdatePath)
+	if err != nil {
+		return "", fmt.Errorf("could not create file at %s: %w", stagedUpdatePath, err)
+	}
+	defer out.Close()
+
+	resp, err := ulm.mirrorClient.Get(ulm.mirrorUrl + fmt.Sprintf("/kolide/%s/%s/%s", binary, ulm.operatingSystem, targetFilename))
+	if err != nil {
+		return stagedUpdatePath, fmt.Errorf("could not make request to download target %s: %w", targetFilename, err)
+	}
+	defer resp.Body.Close()
+
+	_, err = io.Copy(out, resp.Body)
+	if err != nil {
+		return stagedUpdatePath, fmt.Errorf("could not write downloaded target %s to file %s: %w", targetFilename, stagedUpdatePath, err)
+	}
+
+	return stagedUpdatePath, nil
+}
+
+// verifyStagedUpdate checks the downloaded update against the metadata in the TUF repo.
+func (ulm *updateLibraryManager) verifyStagedUpdate(binary string, stagedUpdate string) error {
```

Comment:
This is re-implementing TUF verification? Why not use TUF's Download directly?

Reply:
We can't because we host our metadata on a different server than the binaries, and the go-tuf client implementation doesn't support that (see [here](https://github.com/theupdateframework/go-tuf/blob/master/client/remote_store.go)).

This function uses go-tuf/client's `VerifyDigest` for verification, which does the majority of the work and was the most I was able to pull out.

Reply:
So either we import the SplitStore, or do this?

Reply:
SplitStore only implements it on the server side, so we'd still have to add an equivalent for the client there anyway